src/mirage_net.mli

```diff
@@ -42,42 +43,50 @@ type stats = {
 module type S = sig
 
   type error = private [> Mirage_device.error]
-  (** The type for network errors. *)
+  (** The type for network interface errors. *)
 
   val pp_error: error Fmt.t
   (** [pp_error] is the pretty-printer for errors. *)
 
-  type page_aligned_buffer
-  (** The type for page-aligned memory buffers. *)
-
   type buffer
   (** The type for memory buffers. *)
 
   type macaddr
-  (** The type for unique MAC identifiers for the device. *)
+  (** The type for unique MAC identifiers for the network interface. *)
 
   include Mirage_device.S
 
   val write: t -> buffer -> (unit, error) result io
-  (** [write nf buf] outputs [buf] to netfront [nf]. *)
+  (** [write net buf] outputs [buf] to network interface [net]. *)
 
   val writev: t -> buffer list -> (unit, error) result io
-  (** [writev nf bufs] output a list of buffers to netfront [nf] as a
-      single packet. *)
+  (** [writev net bufs] output a list of buffers to the network interface [net]
+      as a single packet. *)
 
   val listen: t -> (buffer -> unit io) -> (unit, error) result io
-  (** [listen nf fn] is a blocking operation that calls [fn buf] with
+  (** [listen net fn] is a blocking operation that calls [fn buf] with
```

Comment:
Might be interesting to specify here if the `buffer` is supposed to be own by the network interface and might be re-used when the callback is finished or if the consumer can do whatever it wants with the buffer (for instance store it in an external table) because it was copied before calling the callback.

Something like "the lifetime of the buffer is expected to be the managed by the network device, if you want to use the read buffer safely outside of the callback, copy the buffer manually." What do you think ?

Reply:
yes - it's about lifetime and ownership. AFAICT `listen` (i looked into solo5 and unix implementations) actually (a) allocates a buffer (b) reads from the network device (c) creates an asynchronous task and passes the buffer to it. lifetime thereof is until the task finished and ownership is transferred to the task.

Reply:
So reading packets means allocating a lots of page ? This seems sub-optimal. Would probably be better if unix and solo5 implementation were using a pool of pages (and block/discard if no pages are available ?). Should the buffer allocation function be passed as argument as well ?

Reply:
hmm, I spent most of my recent time on the sending path, and figured lifetime/ownership/allocation there. on the receiving path, I'm not yet sure what a sensible way to deal would be:
(a) very simple solution: `listen` allocates initially a single buffer, and does not create any tasks, but instead calls `fn` after it read from the interface in the same task -- only once that `fn` is finished, `read` is called again, reusing the very same page. needs to state that `fn` needs to copy if it wants to retain the data
(b) `listen` uses a buffer pool - pre-allocating N pages initially, then calling read with the next free one, and creating an asynchronous task for each `fn` invocation -- but how should `fn` communicate that the given buffer is not used anymore? xen seems to use a ringbuffer and id interface, where the guest pushes memory into the ringbuffer with identifiers, and the host reports which identifiers are sent and can be safely reused/deallocated..
(c) the `fn` passed to `listen` could be modified to `buffer -> unit`, i.e. not in the io monad. this would allow an (a) solution (but still does not avoid the issue of storing a reference to the buffer anywhere).

my experience from the sending side is that the network device should be responsible for allocating (and it is the one which knows what buffer to pass to the underlying device, i.e. some need it page-aligned, others don't care), plus it knows the MTU. I don't think upper layers should provide allocators downwards.

but somehow the re-using of buffers scares me, we can't verify that any `fn` does not hold a reference to that buffer and reads it at a later stage - which may as well be a security concerns since now your web server can read the buffer used for receiving syslog messages. I'd propose to think a bit more about that, implement several strategies and evaluate the performance. I'll hopefully get around to writeup and push my sending part modifications this weekend.

Reply:
I'm fine to think / improve this later ; all your nice changes in the sending side had make me think on the receiving one :-) But no hurry and this don't need to be tied to that PR at all and can be done later.